internal/server/singleprocess/auth.go

```diff
@@ -57,44 +61,95 @@ var (
+	return context.WithValue(ctx, userKey{}, u)
+}
+
+// userFromContext returns the authenticated user in the request context.
+// This will return nil if the user is not authenticated.
+func (s *service) userFromContext(ctx context.Context) *pb.User {
+	value, ok := ctx.Value(userKey{}).(*pb.User)
+	if !ok && s.superuser {
+		value = &pb.User{Id: DefaultUserId, Username: DefaultUser}
+	}
+
+	return value
+}
+
 // Authenticate implements the server.AuthChecker interface.
 //
 // This checks if the given endpoint should be allowed. This is called during a
 // gRPC request. Effects is some information about the endpoint, at present these are
 // either ["readonly"] or ["mutable"] to indicate if the endpoint will be only reading
 // data or also mutating it.
-func (s *service) Authenticate(ctx context.Context, token, endpoint string, effects []string) error {
+func (s *service) Authenticate(
+	ctx context.Context, token, endpoint string, effects []string,
+) (context.Context, error) {
 	// Ignore unauthenticated endpoints
 	if _, ok := unauthenticatedEndpoints[endpoint]; ok {
-		return nil
+		return nil, nil
 	}
 
 	if token == "" {
-		return status.Errorf(codes.Unauthenticated, "Authorization token is not supplied")
+		return nil, status.Errorf(codes.Unauthenticated, "Authorization token is not supplied")
 	}
 
 	_, body, err := s.decodeToken(token)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	// Token must be a login token to be used for auth
 	login, ok := body.Kind.(*pb.Token_Login_)
 	if !ok || login == nil {
-		return ErrInvalidToken
+		return nil, ErrInvalidToken
 	}
 
 	// If this is an entrypoint token then we can only access entrypoint APIs.
 	if login.Login.Entrypoint != nil && !strings.HasPrefix(endpoint, "Entrypoint") {
-		return status.Errorf(codes.Unauthenticated, "Unauthorized endpoint")
+		return nil, status.Errorf(codes.Unauthenticated, "Unauthorized endpoint")
 	}
 
-	// TODO When we have a user model, this is where you'll check for the user.
-	if login.Login.UserId != DefaultUser {
-		return ErrInvalidToken
+	// Look up the user that this token is for.
+LOOKUP_USER:
+	user, err := s.state.UserGet(&pb.Ref_User{
+		Ref: &pb.Ref_User_Id{
+			Id: &pb.Ref_UserId{Id: login.Login.UserId},
+		},
+	})
+	if status.Code(err) == codes.NotFound {
+		// If we are a legacy token logging into a WP 0.5+ server for the
+		// first time, we need to create the initial Waypoint user. This is
+		// purely a backwards compatibility case that we should drop at some
+		// point.
+		if !body.UnusedLogin || login.Login.UserId != DefaultUserId {
+			return nil, status.Errorf(codes.Unauthenticated,
+				"Pre-Waypoint 0.5 token must be for the default user. This should always "+
+					"be the case so the token is likely corrupt.")
+		}
+
+		// Bootstrap our user
+		if _, err := s.bootstrapUser(); err != nil {
+			// If we're already bootstrapped, give a slightly better error.
```

Comment:
Yes, I thought of this: Two distinct users will never be created because they're created with the same ID. So the race will simply overwrite one with the other. I also felt that given bootstrap users PLUS legacy tokens are going to be so rare, its unlikely to ever cause real issues in any scenario.

Reply:
Ah ah ok. Gotcha sure.